src/qttools/obc/spectral.py

```diff
@@ -166,12 +222,40 @@ def _find_reflected_modes(
 
             ks = -1j * xp.log(ws)
 
+        # replace nan and infs with 0 due to zero eigenvalues
+        dEk_dk = xp.nan_to_num(dEk_dk, nan=0, posinf=0, neginf=0)
+        ks = xp.nan_to_num(ks, nan=0, posinf=0, neginf=0)
+
         # Find eigenvalues that correspond to reflected modes. These are
         # modes that either propagate into the leads or decay away from
         # the system.
-        return ((dEk_dk.real < 0) & (xp.abs(ks.imag) < self.min_decay)) | (
-            (ks.imag < -self.min_decay) & (ks.imag > -self.max_decay)
+        if self.filter_pairwise:
+            # either matched modes with slow enough decay
+            mask_propagating = self._filter_pairwise(dEk_dk, ks)
+            # modes not matched
+            mask_decaying = ~mask_propagating
+        else:
+            # slow enough decay
+            mask_propagating = xp.abs(ks.imag) < self.min_decay
+
+        # fast enough decay
+        mask_decaying = (
+            mask_decaying if self.filter_pairwise else xp.ones_like(dEk_dk, dtype=bool)
+        )
+        mask_decaying &= ks.imag < -self.min_decay
+
```

Comment:
Control flow a little bit weird here, why not something like
```suggestion
        # Determine (matched) modes that decay slow enough to be
        # considered propagating.
        if self.filter_pairwise:
            mask_propagating = self._filter_pairwise(dEk_dk, ks)
            mask_decaying = ~mask_propagating
        else:
            mask_propagating = xp.abs(ks.imag) < self.min_decay
            mask_decaying = xp.ones_like(dEk_dk, dtype=bool)
        
        # Make sure decaying modes decay fast enough.
        mask_decaying &= ks.imag < -self.min_decay

```